packages/engine-core/src/data-proxy/DataProxyEngine.ts

```diff
@@ -219,11 +227,73 @@ export class DataProxyEngine extends Engine {
     }
   }
 
-  // TODO: figure out how to support transactions
-  transaction(): Promise<any> {
-    throw new NotImplementedYetError('Interactive transactions are not yet supported', {
-      clientVersion: this.clientVersion,
-    })
+  /**
+   * Send START, COMMIT, or ROLLBACK to the Query Engine
+   * @param action START, COMMIT, or ROLLBACK
+   * @param headers headers for tracing
+   * @param options to change the default timeouts
+   * @param info transaction information for the QE
+   */
+  async transaction(action: 'start', headers: Tx.TransactionHeaders, options?: Tx.Options): Promise<Tx.Info>
+  async transaction(action: 'commit', headers: Tx.TransactionHeaders, info: Tx.Info): Promise<undefined>
+  async transaction(action: 'rollback', headers: Tx.TransactionHeaders, info: Tx.Info): Promise<undefined>
+  async transaction(action: any, headers: Tx.TransactionHeaders, arg?: any) {
+    await this.start()
+
+    const errorHandler = async (response: RequestResponse) => {
+      if (!response.ok) {
+        const json = await response.json()
+
+        if (json) {
+          throw new Error(`Error in transaction start: ${JSON.stringify(json)}`)
+        } else {
+          throw new Error(`Error in transaction start: ${response.text()}`)
```

Comment:
Ideally, there would be an error code and be documented like the rest of the errors. There a documented on the Data Proxy page in our docs.

Reply:
In which cases can a transaction not start?

Reply:
I would expect errors for transaction start being pretty similar to query request, so the code should use `responseToError` helper and handle them pretty much the same way (including retries and uploading the schema). As for commit and rollback, we should definitely use the same mechanism for error handling and convert user-facing errors to instances of standard Data Proxy errors before throwing them, but maybe we don't have to handle `SchemaMissingError` for those two.